### PR TITLE
Create Extension using codestarts

### LINKED ISSUE #12522: Implement the Create Extension mojo using Codestarts
**Description**
It would be nice if `devtools/maven/src/main/java/io/quarkus/maven/CreateExtensionMojo.java` were implemented using Codestarts instead

**Implementation ideas**
Use Codestarts. Ping @ia3andy for help in quarkusio.zulipchat.com

### PR DESCRIPTION
**Pull request by ia3andy**
Replace (and break) the current `create-extension` mojo using codestarts and with simpler parameters.

The legacy mojo is now `create-extension-legacy`.

In this implementation, it supports creating:
- a Quarkus Core extension from the `quarkus` or `quarkus/extensions` dir
- a (single) Quarkiverse extension using `-DgroupId=io.quarkiverse.[extensionId]`
- a standalone extension from any other dir with another groupId

from the quarkus dir, or any other dir:
```shell
mvn io.quarkus:quarkus-maven-plugin:999-SNAPSHOT:create-extension -N
```
or quarkiverse:
```shell
mvn io.quarkus:quarkus-maven-plugin:999-SNAPSHOT:create-extension -N -DgroupId=io.quarkiverse.aloha
```

You need to also pay a good attention to the updated guides:
- docs/src/main/asciidoc/building-my-first-extension.adoc
- docs/src/main/asciidoc/writing-extensions.adoc

I've added backward compat for the `artifactId` which now is replaced by `extensionId`:
- `artifactId=quarkus-aloha` ->  `extensionId=aloha` and `namespaceId=quarkus-`
- `artifactId=aloha` ->  `extensionId=aloha` and `namespaceId=""`
- `artifactId=something-aloha` ->  `extensionId=something-aloha` and `namespaceId=""`

---

This PR also introduce snapshot testing for generated code `SnapshotTesting` and use it in some of our existing tests:
```java
assertThatMatchSnapshot(testInfo, projectDir, "src/main/java/org/acme/GreetingResource.java")
    .satisfies(checkContains("@Path(\"/hello\")"));
```
and to check a directory tree:
```java
assertThatDirectoryTreeMatchSnapshots(testInfo, projectDir);
```

The snapshots let us review the output of the command and make sure future changes are working as expected.

Snapshots are automatically generated in `src/test/resources/__snapshots__/...` on first run or can be manually updated using `-Dsnap` or `-DupdateSnapshots`.

---

Sorry for the size of this PR:
- it lasted way more than I expected
- it includes the snapshots which introduce new files
- it mostly affect the `create-extension` mojo and code generation tests so the risk is limited

---

Closes #12522

### CODE REVIEW

**Comment by ia3andy** [comment]
@gsmet I added you as reviewer for the guide part and to make sure I included all necessary features

@gastaldi Could you thoroughly check the Quarkiverse mode to make sure such generated extensions are fine for it and maybe create an issue update the doc there when it's out

@aloubyansky this is the PR that may conflict with your new catalog api branch

**Comment by aloubyansky** [comment]
Looks good. I personally would still prefer specifying the artifactId instead of the feature though.

**Comment by ia3andy** [comment]
@aloubyansky If you look closely, I just renamed it, before it was called `artifactIdBase`. I did it because it's not a `base`, it's the `id` of the extension in the current namespace (`quarkus-` by default), I hesitated with `extensionId` but `feature` was already in the code.
```
private static final String FEATURE = "{extension.feature-id}";
```

Maybe I should go all the way in that direction:
```yaml
namespaceId: "quarkus-"
namespaceName: "Quarkus -"
extensionId: "my-ext"
extensionName: "My Ext"
```

then we derivate those to get multiple artifactIds (parent, runtime, deployment and it)

**Review by aloubyansky — APPROVED** [review]

**Comment by ia3andy** [comment]
@aloubyansky ok as discussed on zulip, I've updated the PR using:
```
namespaceId (optional): "quarkus-"
namespaceName (computed from namespaceId by default): "Quarkus - "
extensionId: "my-ext"
extensionName (computed from extensionId by default): "My Ext"
```

Also instead of showing a warning, I've added backward compat for the `artifactId`:
- `artifactId=quarkus-aloha` ->  `extensionId=aloha` and `namespaceId=quarkus-`
- `artifactId=aloha` ->  `extensionId=aloha` and `namespaceId=""`

**Comment by aloubyansky** [comment]
Looks good to me. Thanks!

**Comment by ia3andy** [comment]
Cool! 

I cleaned a bit the `CreateExtension` code, and refreshed the templates from recent quarkiverse changes. This doesn't require a new review..

I will merge as soon as CI passes, we can always fix minor problems in other PRs.

@gastaldi it would be nice if you could check the quarkiverse output:
```shell
mvn io.quarkus:quarkus-maven-plugin:999-SNAPSHOT:create-extension -N -DgroupId=io.quarkiverse.aloha
```


@gsmet let me know if you find probs in the guides.. I think I will update them very soon to make them clearer:
- a guide for "Create you own extension" 
	- Create a Quarkiverse extension
	- Create a Standalone extension
	- Create a Quarkus Core extension
	- Add Unit tests
	- Add Dev mode tests
	- Add integration tests
- a guide for: "Understand extension principles" (with most of the content of `docs/src/main/asciidoc/writing-extensions.adoc`)

**Comment by ia3andy** [comment]
And thanks @aloubyansky, this was a big one! :)

**Comment by ia3andy** [comment]
@gastaldi 😂
image

**Comment by gastaldi** [comment]
@ia3andy lol, yeah, I thought it would be simple but apparently I was wrong. I am testing the Quarkiverse generation now

File: independent-projects/tools/devtools-testing/src/main/java/io/quarkus/devtools/testing/SnapshotTesting.java [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+            final String converted = new String(inputBytes, StandardCharsets.UTF_8);
+            final byte[] outputBytes = converted.getBytes(StandardCharsets.UTF_8);
+            return Arrays.equals(inputBytes, outputBytes);
+        } catch (IOException e) {
+            return false;
+        }
+    }
+
+    private static void deleteExistingSnapshots(String name, Path snapshots) {
+        System.out.println("\n>>>>>> DELETING EXISTING TEST SNAPSHOTS FOR:\n>>>>>> " + name + "\n");
```
Comment (gastaldi) [inline independent-projects/tools/devtools-testing/src/main/java/io/quarkus/devtools/testing/SnapshotTesting.java:226]: Is this System.out necessary or can it use a logger instead?

Comment (ia3andy) [reply independent-projects/tools/devtools-testing/src/main/java/io/quarkus/devtools/testing/SnapshotTesting.java:226]: It's used only in the tests, goal is to make it really obvious, log are not always displayed..

Comment (ia3andy) [reply independent-projects/tools/devtools-testing/src/main/java/io/quarkus/devtools/testing/SnapshotTesting.java:226]: You only see it when using `-Dsnap`

File: independent-projects/tools/devtools-common/src/main/resources/codestarts/quarkus-extension/code/quarkiverse/java/LICENCE [hunk 2]
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
```
Comment (gastaldi) [inline independent-projects/tools/devtools-common/src/main/resources/codestarts/quarkus-extension/code/quarkiverse/java/LICENCE:1]: Rename this file to `LICENSE` (with an `S`)

Comment (ia3andy) [reply independent-projects/tools/devtools-common/src/main/resources/codestarts/quarkus-extension/code/quarkiverse/java/LICENCE:1]: Oupsy :)

Comment (ia3andy) [reply independent-projects/tools/devtools-common/src/main/resources/codestarts/quarkus-extension/code/quarkiverse/java/LICENCE:1]: licence is french 😅 I always mix them up

File: independent-projects/tools/devtools-common/src/main/resources/codestarts/quarkus-extension/code/integration-test/java/integration-test/pom.tpl.qute.xml [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        <artifactId>{it-parent.artifact-id}</artifactId>
+        <version>{it-parent.version}</version>
+        {#if it-parent.relative-path}
+        <relativePath>{it-parent.relative-path}</relativePath>
+        {/if}
+    </parent>
+{/if}
+    <artifactId>{namespace.id}{extension.id}-integration-test</artifactId>
+{#if name.base && name.prefix}
+    <name>{name.prefix}{name.base}{name.segment-delimiter}Integration Test</name>
```
Comment (gastaldi) [inline independent-projects/tools/devtools-common/src/main/resources/codestarts/quarkus-extension/code/integration-test/java/integration-test/pom.tpl.qute.xml:18]: The Quarkiverse extension is not generating a name for the `integration-test` module.

Comment (ia3andy) [reply independent-projects/tools/devtools-common/src/main/resources/codestarts/quarkus-extension/code/integration-test/java/integration-test/pom.tpl.qute.xml:18]: ok this is a bug :) thanks

Comment (ia3andy) [reply independent-projects/tools/devtools-common/src/main/resources/codestarts/quarkus-extension/code/integration-test/java/integration-test/pom.tpl.qute.xml:18]: good catch !

**Comment by gastaldi** [comment]
@ia3andy it looks good, however I noticed it generated the parent pom with the `quarkiverse-parent` 5, but 6 is the latest. Is that fixed somewhere? 

Also, the plugin asks for the extension ID and ignores the `-B` (for batch mode) flag. Can we suggest a default value based on the `groupId`?

**Comment by ia3andy** [comment]
> @ia3andy it looks good, however I noticed it generated the parent pom with the `quarkiverse-parent` 5, but 6 is the latest. Is that fixed somewhere?
> 

Ok I'll update the version to 6, for now it's a constant, we need to make that dynamic in a future PR ;)

> Also, the plugin asks for the extension ID and ignores the `-B` (for batch mode) flag. Can we suggest a default value based on the `groupId`?

Yeah, I'll update ;)

File: independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/commands/CreateExtension.java [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+    public static final String DEFAULT_BOM_VERSION = "${quarkus.version}";
+    public static final String DEFAULT_VERSION = "0.0.1-SNAPSHOT";
+
+    public static final String DEFAULT_CORE_NAMESPACE_ID = "quarkus-";
+
+    public static final String DEFAULT_EXTERNAL_NAMESPACE_ID = "";
+
+    public static final String DEFAULT_QUARKIVERSE_PARENT_GROUP_ID = "io.quarkiverse";
+    public static final String DEFAULT_QUARKIVERSE_PARENT_ARTIFACT_ID = "quarkiverse-parent";
+    public static final String DEFAULT_QUARKIVERSE_PARENT_VERSION = "5";
```
Comment (gastaldi) [inline independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/commands/CreateExtension.java:52]: Hint: these properties can be moved as getters inside the `LayoutType` enum above

Comment (ia3andy) [reply independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/commands/CreateExtension.java:52]: Good hint 😇

**Comment by gastaldi** [comment]
> Ok I'll update the version to 6, for now it's a constant, we need to make that dynamic in a future PR ;)

The best way to do that is with an instance of `io.quarkus.bootstrap.resolver.maven.MavenArtifactResolver` (which you already have in the classpath)

**Comment by ia3andy** [comment]
> > Ok I'll update the version to 6, for now it's a constant, we need to make that dynamic in a future PR ;)
> 
> The best way to do that is with an instance of `io.quarkus.bootstrap.resolver.maven.MavenArtifactResolver` (which you already have in the classpath)

I am sure you will do it as soon as this one gets merged :)

**Comment by gastaldi** [comment]
> I am sure you will do it as soon as this one gets merged :)

I wouldn't be so sure it would happen soon after that, but I'll add to my TODO list (unless you implement that first) 😉

**Comment by ia3andy** [comment]
ok @gastaldi, I fixed it all and only `groupId` is needed for quarkiverse :):

interactive mode:
```
[INFO] Scanning for projects...
[INFO]
[INFO] ------------------< org.apache.maven:standalone-pom >-------------------
[INFO] Building Maven Stub Project (No POM) 1
[INFO] --------------------------------[ pom ]---------------------------------
[INFO]
[INFO] --- quarkus-maven-plugin:999-SNAPSHOT:create-extension (default-cli) @ standalone-pom ---
Set the extension groupId [org.acme]: io.quarkiverse.stella

Detected layout type is 'quarkiverse'
Generated runtime artifactId is 'quarkus-stella'


applying codestarts...
🔠 java
🧰 maven
🗃 quarkus-extension
🛠 git
🐒 devmode-test
🐒 extension-base
🐒 integration-test
🐒 quarkiverse
🐒 unit-test

-----------
👍 extension has been successfully generated in:
--> /Users/ia3andy/workspace/redhat/quarkus/demo/quarkus-stella
-----------
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  10.049 s
[INFO] Finished at: 2021-01-29T16:01:31+01:00
[INFO] ------------------------------------------------------------------------
```

and manual mode:
```
mvn io.quarkus:quarkus-maven-plugin:999-SNAPSHOT:create-extension -N -DgroupId=io.quarkiverse.quarkila
[INFO] Scanning for projects...
[INFO]
[INFO] ------------------< org.apache.maven:standalone-pom >-------------------
[INFO] Building Maven Stub Project (No POM) 1
[INFO] --------------------------------[ pom ]---------------------------------
[INFO]
[INFO] --- quarkus-maven-plugin:999-SNAPSHOT:create-extension (default-cli) @ standalone-pom ---

Detected layout type is 'quarkiverse'
Generated runtime artifactId is 'quarkus-quarkila'


applying codestarts...
🔠 java
🧰 maven
🗃 quarkus-extension
🛠 git
🐒 devmode-test
🐒 extension-base
🐒 integration-test
🐒 quarkiverse
🐒 unit-test

-----------
👍 extension has been successfully generated in:
--> /Users/ia3andy/workspace/redhat/quarkus/demo/quarkus-quarkila
-----------
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  1.709 s
[INFO] Finished at: 2021-01-29T15:58:59+01:00
[INFO] ------------------------------------------------------------------------
```

**Pull request merged**